state/optimistic_syncing/pipeline.go

```diff
@@ -0,0 +1,474 @@
+			if shouldAbandon := p.updateState(update); shouldAbandon {
+				p.broadcastStateUpdate()
+				cancelFunc()
+			} else {
+				// Trigger state check
+				p.stateNotifier.Notify()
+			}
+		}
+	}
+}
+
+// broadcastStateUpdate sends a state update to all children pipelines.
+// The update includes the current pipeline state and whether it descends
+// from the latest persisted result.
+func (p *Pipeline) broadcastStateUpdate() {
+	if len(p.childrenUpdateChans) == 0 {
+		return
+	}
+
+	p.mu.RLock()
+	update := StateUpdate{
+		DescendsFromLastPersistedSealed: p.descendsFromSealed,
+		ParentState:                     p.state,
+	}
+	channels := p.childrenUpdateChans
+	p.mu.RUnlock()
+
+	for _, ch := range channels {
+		ch <- update
+	}
+}
+
+// processCurrentState handles the current state and transitions to the next state if possible.
+// It returns false when a terminal state is reached (StateComplete or StateCanceled), true otherwise.
+// Returns an error if any processing step fails.
+func (p *Pipeline) processCurrentState(ctx context.Context) (bool, error) {
+	currentState := p.GetState()
+
+	switch currentState {
+	case StateReady:
+		return p.processReady(), nil
+	case StateDownloading:
+		return p.processDownloading(ctx)
+	case StateIndexing:
+		return p.processIndexing(ctx)
+	case StateWaitingPersist:
+		return p.processWaitingPersist(), nil
+	case StatePersisting:
+		return p.processPersisting(ctx)
+	case StateComplete, StateCanceled:
+		// Terminal states
+		return false, nil
+	default:
+		return false, fmt.Errorf("%w: %s", ErrInvalidState, currentState.String())
+	}
+}
+
+// transitionTo transitions the pipeline to the given state and broadcasts
+// the state change to children pipelines.
+func (p *Pipeline) transitionTo(newState State) {
+	p.mu.Lock()
+	oldState := p.state
+	p.state = newState
+	p.mu.Unlock()
+
+	p.logger.Info().
+		Str("oldState", oldState.String()).
+		Str("newState", newState.String()).
+		Str("executionResultID", p.executionResultID).
+		Str("blockID", p.blockID).
+		Msg("pipeline state transition")
+
+	// Broadcast state update to children
+	p.broadcastStateUpdate()
+
+	// Trigger state check in case we can immediately transition again
+	if newState != StateComplete && newState != StateCanceled {
+		p.stateNotifier.Notify()
+	}
+}
+
+// processReady handles the Ready state and transitions to StateDownloading if possible.
+// Returns true to continue processing, false if a terminal state was reached.
+func (p *Pipeline) processReady() bool {
+	if p.canStartDownloading() {
+		p.transitionTo(StateDownloading)
+		return true
+	}
+	return true
+}
+
+// processDownloading handles the Downloading state.
+// It executes the download function and transitions to StateIndexing if successful.
+// Returns true to continue processing, false if a terminal state was reached.
+// Returns an error if the download step fails.
+func (p *Pipeline) processDownloading(ctx context.Context) (bool, error) {
+	p.logger.Info().
```

Comment:
we should manage how many new log messages are added at info level since they are logged for every block.

as mentioned before, I think all of the "starting" messages should be debug level, and completed messages can be info